### ISSUE TITLE
Receive hover events on i3bar 

### ISSUE THREAD

**Issue by edi9999** [i3/i3#3972]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

It is possible to retrieve click events like this : 

```
{
 "name": "ethernet",
 "instance": "eth0",
 "button": 1,
 "modifiers": ["Shift", "Mod1"],
 "x": 1320,
 "y": 1400,
 "relative_x": 12,
 "relative_y": 8,
 "width": 50,
 "height": 22
}
```

## Desired Behavior

I would like to also retreieve hover events (when the mouse goes over a block, ideally only after a duration of 500ms, because we don't wan to trigger all over events if you just went on the block but you were still moving the mouse).

This could maybe use : "button": null, "hover": true, eg : 

{
 "name": "ethernet",
 "instance": "eth0",
 "button": null,
 "hover": true, 
 "modifiers": ["Shift", "Mod1"],
 "x": 1320,
 "y": 1400,
 "relative_x": 12,
 "relative_y": 8,
 "width": 50,
 "height": 22
}

**Comment by Airblader** [comment]
Thanks for the issue! Could you share your use-cases on what exactly you want to do on hover?

> I would like to also retreieve hover events […] ideally only after a duration of 500ms

One of my two concerns with this feature is exactly this: we definitely don't want to spam events, but we also cannot debounce them to a fixed value as this will quickly become unsuitable for people. Which leaves configuration as the only other choice, but for any new configuration directive we need to weigh the benefit against the cost.

My other concern is compatibility. I would assume there's plenty of scripts out there which don't check for the value of `button` which would then all trigger on hover rather than click. Ideally I would like to avoid this, particularly I think the feature is not "big enough" to justify risking breakage of existing setups.

**Comment by edi9999** [comment]
My idea would be to either : 

* change how the widget is shown (for example, if using a time widget, switch from showing only the time (by default) to show the date + time on hover).

* add some window above the i3bar to show some detailed information about this widget (this would be developped in the user script itself), for example on hover on the CPU stat widget, show the 3 processes that use most CPU.

The feature actually requires not just hover event, but also mouseout events, because the window needs to close when you move the mouse out of the hover region.

I understand your concern about breaking user-scripts, maybe this could be an opt-in feature where you would specify in your i3config : 

```
bar {
        font pango:monospace 9
        ...
        hoverevents: yes
}
```

**Comment by orestisfl** [comment]
> My other concern is compatibility. I would assume there's plenty of scripts out there which don't check for the value of `button` which would then all trigger on hover rather than click. Ideally I would like to avoid this, particularly I think the feature is not "big enough" to justify risking breakage of existing setups.

We can make `"hover_events"` an opt-in configuration option like `"click_events"`:

```
{ "version": 1, "stop_signal": 10, "cont_signal": 12, "click_events": true }
```

**Comment by edi9999** [comment]
Yes that sounds good to me !

**Comment by eleksir** [comment]
Usecase?
Tooltip (kind of). Say, i want to implement "launch buttons". On launch buttons themseves i just put 1 first letter of app name, ex. T (for telegram), S (for slack), T (for thunderbird), T (for terminal/urxvt). And so on, many of them. I hoover mouse on button and see what it is.

Extended info. Let's imagine memory information: in bar i put percentage on used mem and free swap, on mouse hoover, i write exact values in megabytes.

dmenu in i3bar. Sounds like insane idea. Input text field also required for this and ability to capture input for this input text field.

dunno if it really be useful for someone besides me...

**Comment by MaxVerevkin** [comment]
I don't know if it was discussed before, but what about extending the protocol to accept a "hover" or "tooltip" value which will be show on mouse hover in a small window above the bar? And since it will be rendered in a separate window, it can contain multiple lines.
```json
{
    "full_text":"cpu: 20%",
    "tooltip":"core 1: 5%\ncore 2: 35%"
}
```

**Comment by Airblader** [comment]
@MaxVerevkin That'd be a very specific feature built on top of i3bar. First it'll be the text, then the position of the tooltip, the color of the text, the color of the tooltip's background, its font, the duration to display it, a delay for how quickly to display it, …

I don't think this is something we'd want to do – offering hover events to the user instead would be a much more generic mechanism that allows more use cases and better control.

**Comment by edi9999** [comment]
Yes, making the tooltip part of i3bar would be quite complex, I think just doing the event and letting the user handle it would be workable however.

**Comment by MaxVerevkin** [comment]
> offering hover events to the user instead would be a much more generic mechanism that allows more use cases and better control.

There is a limitation though: this event cannot be used to show a tooltip on wayland because clients don't have access to global coordinates. Obviously this is not a problem for i3bar, but the same protocol is used by swaybar and other bars.

I don't know what i3's position is when it comes to the protocol being used by other bars, so maybe this is irrelevant.

**Comment by Airblader** [comment]
I don't think we are worried about other display servers here.
